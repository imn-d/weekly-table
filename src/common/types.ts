import React from 'react';

/**
 * Input props
 * @property parentRef - parent reference or component size
 * @property blockColors - using time block colors, see BlockColorsProps
 * @property headerHeightProp - height of columns header
 * @property helperWidthProp - width of left time column
 * @property bottomHeightProp - height of last rows
 * @property baseZIndex - base z-index
 * @property timeframe - minutes count per 1 cell
 * @property columns - days of week list
 * @property rows - time descriptions list
 * @property mouseSpeed - moving-resizing const
 * @property requiredTZOffset - used timezone
 * @property defaultValue - initial scheduler value
 * @property onChange - callback, schedules return in UTC+0
 */
export interface SchedulerInputProps {
  parentRef: React.RefObject<HTMLDivElement>;
  blockColors?: BlockColorsProps;
  headerHeightProp?: number;
  helperWidthProp?: number;
  bottomHeightProp?: number;
  baseZIndex?: number;
  timeframe?: number;
  columns?: SchedulerColumnsProps[];
  rows?: string[];
  mouseSpeed?: number;
  requiredTZOffset?: number;
  defaultValue?: ScheduleGroup[];
  onChange?: (groups: ScheduleGroup[]) => void;
}

/**
 * Internal props
 * @see SchedulerInputProps
 */
export interface SchedulerProps {
  parentRef: React.RefObject<HTMLDivElement>;
  blockColors: BlockColorsProps;
  headerHeightProp: number;
  helperWidthProp: number;
  bottomHeightProp: number;
  baseZIndex: number;
  timeframe: number;
  columns: SchedulerColumnsProps[];
  rows: string[];
  mouseSpeed: number;
  requiredTZOffset?: number;
  defaultValue?: ScheduleGroup[];
  onChange?: (groups: ScheduleGroup[]) => void;
}

/**
 * Block colors props
 * @property common - common color
 * @property temp - temp block color
 * @property draw - drawing block color
 * @property hover - hover block color
 */
export interface BlockColorsProps {
  common: string;
  temp: string;
  draw: string;
  hover: string;
}

/**
 * Days of week props
 * @property weight - column weight, calculates mask
 * @property full - full display name
 * @property short -  short display name
 */
export interface SchedulerColumnsProps {
  weight: number;
  full: string;
  short: string;
}

/**
 * Schedule props
 * @property startTime - start time in ms
 * @property endTime - end time in ms
 * @property mask - group mask
 */
export interface ScheduleGroup {
  startTime: number;
  endTime: number;
  mask: number;
}

/**
 * Cell props
 * @property position - relative coordinates x,y,w,h
 * @property row - timeframe count in day
 * @property column - day of week
 * @property isSelected - was selected or not
 */
export interface Cell {
  position: {
    x: number;
    y: number;
    w: number;
    h: number;
  };
  row: number;
  column: number;
  isSelected: boolean;
}

/**
 * Virtual cells provider props
 * @property offsetWidth - width of parent component
 * @property offsetHeight - height of parent component
 * @property widthStep - column width
 * @property heightStep - cell height
 * @property millisPerPixel - ms count per pixel
 * @property cells - virtual cells array
 * @property setCells - virtual cells array setter
 * @property msTime - ms count in timeframe
 * @property maxCellIndex - last column cell index
 */
export interface CellsProps {
  offsetWidth: number;
  offsetHeight: number;
  widthStep: number;
  heightStep: number;
  millisPerPixel: number;
  cells: Cell[];
  setCells: (update: Cell[]) => void;
  msTime: number;
  maxCellIndex: number;
}

/**
 * Drawing provider props
 * @property canvasRef - canvas object reference
 * @property startDrawing - callback for starting draw action
 * @property finishDrawing - callback for ending draw action
 * @property draw -  callback for drawing action
 * @property isDrawing - drawing property
 */
export interface CanvasProps {
  canvasRef: React.RefObject<HTMLCanvasElement>;
  startDrawing: ({ nativeEvent }: React.MouseEvent) => void;
  finishDrawing: ({ nativeEvent }: React.MouseEvent) => void;
  draw: ({ nativeEvent }: React.MouseEvent) => void;
  isDrawing: boolean;
}

/**
 * Pointer lock provider props
 * @property moveRef - time block move area reference
 * @property resBotRef - time block resize bottom area reference
 * @property resTopRef - time block resize top area reference
 * @property draw - callback for resizing/moving actions
 * @property isLocking - resizing/moving action property
 * @property movement - mouse relative positions
 * @property action - action type (resize_bot,resize_top,move)
 * @property startDrawing - callback for starting resizing/moving action
 * @property finishDrawing - callback for ending resizing/moving action
 */
export interface PointerLockProps {
  moveRef: React.RefObject<HTMLDivElement>;
  resBotRef: React.RefObject<HTMLDivElement>;
  resTopRef: React.RefObject<HTMLDivElement>;
  draw: ({ nativeEvent }: React.MouseEvent) => void;
  isLocking: boolean;
  movement: number;
  action: number;
  startDrawing: (action: number) => void;
  finishDrawing: () => void;
}

/**
 * Time block props
 * @property id - generated string
 * @property top - top relative position
 * @property left -  left relative position
 * @property width - time block width
 * @property height - time block height
 * @property startTime - start time in ms
 * @property endTime - end time in ms
 * @property column - day of week
 * @property realStartTime - additional prop for block actions
 * @property realEndTime - additional prop for block actions
 * @property isTemp - temporary block property
 */
export interface TimeBlock {
  id: string;
  top: number;
  left: number;
  width: number;
  height: number;
  startTime: number;
  endTime: number;
  column: number;
  realStartTime: number;
  realEndTime: number;
  isTemp: boolean;
}

/**
 * Time block provider props
 * @property blocks - common time blocks array
 * @property preview - temp time blocks array
 * @property setTopPosition - callback-resizer for top positions
 * @property setBotPosition - callback-resizer for bottom positions
 * @property setBlocks - time blocks setter
 * @property handleHistory - saving time block history
 * @property undoHistory - restoring time block history
 */
export interface TimeBlockProps {
  blocks: TimeBlock[];
  preview: TimeBlock[];
  setTopPosition: (block: TimeBlock, withHeight: boolean) => void;
  setBotPosition: (block: TimeBlock) => void;
  setBlocks: React.Dispatch<React.SetStateAction<TimeBlock[]>>;
  handleHistory: (input: TimeBlock[]) => void;
  undoHistory: () => void;
}

/**
 * Popup props
 * @property id - popup id
 * @property block - time block for popup
 */
export interface PopupType {
  id: string;
  block: TimeBlock;
}

/**
 * Popup provider props
 * @property popup - popup
 * @property showPopup - showing popup
 * @property deleteBlock - deleting time block
 * @property hidePopup - hiding popup
 * @property isAllowing - timeRange check
 */
export interface PopupProps {
  popup: PopupType | undefined;
  showPopup: (blockId: string) => void;
  deleteBlock: (blockId: string) => void;
  hidePopup: () => void;
  isAllowing: boolean;
}

/**
 * Position props for custom time component
 * @property top - relative top position
 * @property left - relative bottom position
 */
export interface Position {
  top: number;
  left: number;
}

/**
 * Time props for custom time component (ms)
 * @property startTime - ms start time
 * @property endTime - ms end time
 */
export interface Time {
  startTime: number;
  endTime: number;
}

/**
 * Displaying time props for custom time component (string)
 * @property startTime - start time as string
 * @property endTime - end time as string
 */
export interface DisplayTime {
  startTime: string;
  endTime: string;
}

/**
 * Custom time provider props
 * @property isEditing - active editing property
 * @property updateBlock - applying time for time block
 * @property setEditing - active editing setter
 * @property handleTime - saving new time
 * @property position - timers relative position
 * @property handleObject - time block handling to provider
 * @property resetTime - resetting time
 * @property isError - error text
 * @property displayTime - visible time
 */
export interface CustomTimeProps {
  isEditing: boolean;
  updateBlock: () => boolean;
  setEditing: React.Dispatch<React.SetStateAction<boolean>>;
  handleTime: (newTime: number, type: boolean) => void;
  position: Position | undefined;
  handleObject: (block: TimeBlock) => void;
  resetTime: () => void;
  isError: string | undefined;
  displayTime: DisplayTime;
}
