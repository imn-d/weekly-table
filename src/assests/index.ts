export * from './CancelIcon';
export * from './DeleteIcon';
export * from './EditIcon';
export * from './DragIcon';
export * from './CheckIcon';
