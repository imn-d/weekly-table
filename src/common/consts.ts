import { SchedulerColumnsProps } from './types';

/**
 * Default used columns descriptions
 * [SchedulerColumnsProps]{@link SchedulerColumnsProps}
 */
export const schedulerColumns: SchedulerColumnsProps[] = [
  { weight: 1, full: 'Monday', short: 'Mon' },
  { weight: 2, full: 'Tuesday', short: 'Tue' },
  { weight: 4, full: 'Wednesday', short: 'Wed' },
  { weight: 8, full: 'Thursday', short: 'Thu' },
  { weight: 16, full: 'Friday', short: 'Fri' },
  { weight: 32, full: 'Saturday', short: 'Sat' },
  { weight: 64, full: 'Sunday', short: 'Sun' },
];

/**
 * Default used time rows descriptions
 */
export const schedulerRows: string[] = [
  '01:00',
  '04:00',
  '07:00',
  '10:00',
  '13:00',
  '17:00',
  '20:00',
  '23:00',
];

/**
 * Describing time block moving direction
 */
export const DIR = {
  TOP: false,
  BOTTOM: true,
};

/**
 * Describing time block actions
 */
export const ACTION = {
  RESIZE_TOP: 0,
  RESIZE_BOT: 1,
  MOVE: 2,
};

/**
 * Describing which time was changes, using for custom time setter
 */
export const TIME = {
  START: true,
  END: false,
};

/**
 * ms count per 1 day
 */
export const HOUR_24 = 86400000;
