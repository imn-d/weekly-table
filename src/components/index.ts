export * from './Background';
export * from './Popup';
export * from './TimeBlockCore';
export * from './CustomTime';
export * from './SchedulerArea';
export * from './SchedulerHeader';
export * from './SchedulerHelper';
export * from './SchedulerLayout';
