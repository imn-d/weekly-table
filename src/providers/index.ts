export * from './CanvasProvider';
export * from './CellsProvider';
export * from './CustomTimeProvider';
export * from './PointerLockProvider';
export * from './PopupProvider';
export * from './Scheduler';
export * from './TimeBlockProvider';
