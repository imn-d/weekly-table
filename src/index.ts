import {
  BlockColorsProps,
  ScheduleGroup,
  SchedulerColumnsProps,
  SchedulerInputProps,
  schedulerColumns,
  schedulerRows,
} from './common';
import Scheduler from './providers/Scheduler';

export default Scheduler;
export { schedulerColumns, schedulerRows };
export type {
  SchedulerColumnsProps,
  SchedulerInputProps,
  ScheduleGroup,
  BlockColorsProps,
};
