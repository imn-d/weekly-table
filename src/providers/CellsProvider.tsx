import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { Cell, CellsProps, HOUR_24 } from '../common';
import { useScheduler } from './Scheduler';

const CellsContext = createContext<CellsProps>({} as CellsProps);
/**
 *  Virtual cells provider context hook
 *  @returns [CellsProps]{@link CellsProps}
 */
export const useCells = (): CellsProps => useContext(CellsContext);

/**
 * Virtual cells provider
 */
const CellsProvider: FC = ({ children }) => {
  const {
    parentRef,
    timeframe,
    columns,
    helperWidthProp,
    headerHeightProp,
    bottomHeightProp,
  } = useScheduler();

  /**
   * Main context state, setting by useEffect
   */
  const [cellsProps, setCellsProps] = useState<CellsProps>({} as CellsProps);

  /**
   * Main context updater
   */
  const setCells = useCallback(
    (update: Cell[]) => {
      setCellsProps({ ...cellsProps, cells: update });
    },
    [cellsProps],
  );

  /**
   * Ms count per timeframe
   */
  const msTime = useMemo(() => {
    return timeframe * 60 * 1000;
  }, [timeframe]);

  /**
   * Calculates context values and build virtual cells area
   */
  useEffect(() => {
    const parent = parentRef?.current;
    if (!parent) return;
    const widthStep = (parent.offsetWidth - helperWidthProp) / columns?.length;
    const areaHeight =
      parent.offsetHeight - headerHeightProp - bottomHeightProp;
    const heightStep = areaHeight / (1440 / timeframe);
    const millisPerPixel = HOUR_24 / areaHeight;
    const maxCellIndex = areaHeight / heightStep - 1;
    const cells = new Array<Cell>();
    let widthPos = helperWidthProp;
    for (let i = 0; i < columns?.length; i++) {
      let heightPos = headerHeightProp;
      for (let j = 0; j < areaHeight / heightStep; j++) {
        cells.push({
          isSelected: false,
          column: i,
          row: j,
          position: {
            x: widthPos,
            y: heightPos,
            w: widthPos + widthStep,
            h: heightPos + heightStep,
          },
        } as Cell);
        heightPos += heightStep;
      }
      widthPos += widthStep;
    }

    setCellsProps({
      widthStep,
      heightStep,
      millisPerPixel,
      cells,
      maxCellIndex,
      offsetWidth: parent.offsetWidth,
      offsetHeight: parent.offsetHeight,
    } as CellsProps);
  }, [
    parentRef?.current?.offsetWidth,
    parentRef?.current?.offsetHeight,
    columns?.length,
    bottomHeightProp,
    headerHeightProp,
    helperWidthProp,
    timeframe,
    parentRef,
  ]);

  return (
    <CellsContext.Provider
      value={{
        ...cellsProps,
        setCells: setCells,
        msTime,
      }}
    >
      {children}
    </CellsContext.Provider>
  );
};

export default CellsProvider;
