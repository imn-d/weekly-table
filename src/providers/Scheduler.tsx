import React, {
  FC,
  createContext,
  useContext,
  useEffect,
  useState,
} from 'react';

import {
  BlockColorsProps,
  SchedulerInputProps,
  SchedulerProps,
  schedulerColumns,
  schedulerRows,
} from '../common';
import { Background, CustomTime, Popup, SchedulerLayout } from '../components';
import CanvasProvider from './CanvasProvider';
import CellsProvider from './CellsProvider';
import CustomTimeProvider from './CustomTimeProvider';
import PointerLockProvider from './PointerLockProvider';
import PopupProvider from './PopupProvider';
import TimeBlockProvider from './TimeBlockProvider';

const SchedulerContext = createContext<SchedulerProps>({} as SchedulerProps);
/**
 *  Scheduler provider context hook
 *  @returns [SchedulerProps]{@link SchedulerProps}
 */
export const useScheduler = (): SchedulerProps => useContext(SchedulerContext);

/**
 * Root scheduler components
 * @param props{@link SchedulerInputProps} initial props
 */
const Scheduler: FC<SchedulerInputProps> = (props: SchedulerInputProps) => {
  const [update, setUpdate] = useState<number>(0);

  useEffect(() => {
    window.addEventListener('resize', () => setUpdate((update) => update + 1));
    return () => {
      window.removeEventListener('resize', () =>
        setUpdate((update) => update + 1),
      );
    };
  }, []);

  return (
    <SchedulerContext.Provider
      value={{
        parentRef: props.parentRef,
        blockColors: props?.blockColors
          ? props.blockColors
          : ({
              common: '#ff5722',
              temp: '#c6a700',
              draw: '#ff8a50',
              hover: '#ff3d00',
            } as BlockColorsProps),
        headerHeightProp: props?.headerHeightProp ? props.headerHeightProp : 80,
        helperWidthProp: props?.helperWidthProp ? props.helperWidthProp : 80,
        bottomHeightProp: props?.bottomHeightProp ? props.bottomHeightProp : 20,
        baseZIndex: props?.baseZIndex ? props.baseZIndex : 0,
        timeframe: props?.timeframe ? props.timeframe : 60,
        columns: props?.columns ? props.columns : schedulerColumns,
        rows: props?.rows ? props.rows : schedulerRows,
        requiredTZOffset: props?.requiredTZOffset,
        mouseSpeed: props?.mouseSpeed ? props.mouseSpeed : 3,
        defaultValue: props?.defaultValue,
        onChange: props?.onChange,
      }}
    >
      <CellsProvider>
        <CanvasProvider>
          <PointerLockProvider>
            <TimeBlockProvider>
              <PopupProvider>
                <CustomTimeProvider>
                  <Background />
                  <SchedulerLayout />
                  <Popup />
                  <CustomTime />
                </CustomTimeProvider>
              </PopupProvider>
            </TimeBlockProvider>
          </PointerLockProvider>
        </CanvasProvider>
      </CellsProvider>
    </SchedulerContext.Provider>
  );
};

export default Scheduler;
