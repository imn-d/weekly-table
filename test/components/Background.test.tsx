import { mount } from 'enzyme';
import React from 'react';

import { CellsProps, SchedulerProps } from '../../src/common';
import { Background } from '../../src/components';
import * as uc from '../../src/providers/CellsProvider';
import * as us from '../../src/providers/Scheduler';

describe('Background Component', () => {
  const useSchedulerMock = jest.spyOn(us, 'useScheduler');
  const useCellMock = jest.spyOn(uc, 'useCells');

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Base Render', () => {
    useSchedulerMock.mockReturnValueOnce({
      helperWidthProp: 80,
      headerHeightProp: 80,
      baseZIndex: 0,
      bottomHeightProp: 20,
    } as SchedulerProps);

    useCellMock.mockReturnValueOnce({
      widthStep: 20,
      heightStep: 10,
      offsetHeight: 100,
      offsetWidth: 200,
    } as CellsProps);

    const wrapper = mount(<Background />);

    expect(useSchedulerMock).toBeCalledTimes(1);
    expect(useCellMock).toBeCalledTimes(1);

    expect(wrapper).toMatchSnapshot('baseRender');
    wrapper.unmount();
  });
});
