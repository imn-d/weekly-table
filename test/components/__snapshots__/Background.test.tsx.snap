// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Background Component Base Render: baseRender 1`] = `
<Memo()>
  <canvas
    className="scheduler-background-canvas"
    id="backgroundCanvas"
    style={
      Object {
        "position": "absolute",
        "zIndex": 1,
      }
    }
  />
</Memo()>
`;
